walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,95 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Comparator;
+
+//PROCESS: Iterate over the events given checking if there are attendees
+//of the meeting request. If there are, put the time (start-end) of the event in a list.
+//At the end, search the empty spaces in the list and add the possible
+//options to the 'options' list
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    //List I'll be returning
+    List<TimeRange> options = new ArrayList();
+    //List where I'll store occupied time slots
+    List<TimeRange> occupied = new ArrayList();
+
+    //SPECIAL CASE: Duration longer than the day = Return empty list
+    if(request.getDuration() > TimeRange.END_OF_DAY){
+        return options;
+    }
+
+    //SPECIAL CASES: No attendees and/or no events = Whole day available
+    if(request.getAttendees().isEmpty() || events.isEmpty()){
+        options.add(TimeRange.WHOLE_DAY);
+        return options;
+    }
+
+    //Checks if the given events contain at least one of the meeting's attendees
+    for(Event event : events) {
+        if(containsAttendee(event, request.getAttendees())){
+            occupied.add(event.getWhen());
+        }
+    }
+
+    //SPECIAL CASE: If occupied is empty at this point, the meeting can be at any time of the day, so return the whole day available
+    if(occupied.isEmpty()){
+        options.add(TimeRange.WHOLE_DAY);
+        return options;
+    }
+
+    //Sorting the occupied time slots by starting time from earliest to latest
+    Comparator<TimeRange> comp = new SortByStartTime();
+    Collections.sort(occupied, comp);
+
+    // |---------|
+    //      |-------|
+    // |------------|
+
+    //Handling overlapping/nested time slots
+    for(int i=0; i<occupied.size()-1; i++){
+        //If 2 time slots overlap, replace it with one time slot that covers both
+        if(occupied.get(i).overlaps(occupied.get(i+1))){
+            if(occupied.get(i+1).end() > occupied.get(i).end()){
+                //New duration = 1st time slot + 2nd time slot - the part where they overlap
+                occupied.set(i, TimeRange.fromStartDuration(occupied.get(i).start(), occupied.get(i).duration() + occupied.get(i+1).duration() - (occupied.get(i).end() - occupied.get(i+1).start())));
```

Comment:
A simpler way would be to use the TimeRange#fromStartEnd, use the start time of the first time slot, and use the maximum of the end times of the first and second time slots. But this works too.

Reply:
Oh, I didn't think of it at the time, but it's much simpler! That would avoid that super long (and confusing at first) line